README.md
# DHCP Python

Version 0.1.4

A Python implementation of a DHCP client and the tools to manipulate DHCP packets. Includes:

1. A parser of DHCP packets, returning Python objects
2. Supports for all DHCP options in RFC 2132
3. A rudimentary DHCP client

## Installation

`pip install dhcppython`

## Requirements

* Python 3.8.0 or higher

**NOTE: This has been tested on Ubuntu 18.04 and Windows WSL. May or may not work on other platforms.**

## The Packet Parser

Two files contribute to the packet parsing: `dhcppython.packet` and `dhcppython.options`. For most operations only `dhcppython.packet` will be required.

### dhcppython.packet

The main class in `dhcppython.packet` is the `DHCPPacket`. The `DHCPPacket` class contains multiple constructors for parsing and constructing DHCP packets. 

#### Converting a packet in wireformat to a Python object

Given a DHCP packet in `bytes` format (such as what you would get from reading a DHCP packet straight from a socket) a DHCPPacket object can be instantiated by calling the `from_bytes` and supplying the bytes.

```python
>>> pkt = dhcppython.packet.DHCPPacket.from_bytes(b'\x01\x01\x06\x00\xea\xbe\xc3\x97\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x8cE\x00E\x12\t\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00c\x82Sc5\x01\x01=\x07\x01\x8cE\x00E\x12\t9\x02\x05\xdc<\x0eandroid-dhcp-9\x0c\tGalaxy-S97\n\x01\x03\x06\x0f\x1a\x1c3:;+\xff')
>>> pkt
DHCPPacket(op='BOOTREQUEST', htype='ETHERNET', hlen=6, hops=0, xid=3938370455, secs=1, flags=0, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('0.0.0.0'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='8C:45:00:45:12:09', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x01'), ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00E\x12\t'), MaxDHCPMessageSize(code=57, length=2, data=b'\x05\xdc'), VendorClassIdentifier(code=60, length=14, data=b'android-dhcp-9'), Hostname(code=12, length=9, data=b'Galaxy-S9'), ParameterRequestList(code=55, length=10, data=b'\x01\x03\x06\x0f\x1a\x1c3:;+'), End(code=255, length=0, data=b'')]))
```

#### Converting a DHCPPacket object to wireformat

Given a DHCPPacket object you can easily output the corresponding DHCP packet in wireformat by accessing the `asbytes` attribute of the object.

```python
>>> pkt.asbytes
b'\x01\x01\x06\x00\xea\xbe\xc3\x97\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x8cE\x00E\x12\t\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00c\x82Sc5\x01\x01=\x07\x01\x8cE\x00E\x12\t9\x02\x05\xdc<\x0eandroid-dhcp-9\x0c\tGalaxy-S97\n\x01\x03\x06\x0f\x1a\x1c3:;+\xff'
```

This bytes output is suitable for sending over a socket to a DHCP server.

#### Other Constructors of the DHCPPacket Class

* The default, low level, constructor (not recommended):

```python
>>> pkt = dhcppython.packet.DHCPPacket(op="BOOTREQUEST", htype="ETHERNET", hlen=6, hops=0, xid=123456, secs=0, flags=0, ciaddr=ipaddress.IPv4Address(0), yiaddr=ipaddress.IPv4Address(0), siaddr=ipaddress.IPv4Address(0), giaddr=ipaddress.IPv4Address(0), chaddr="DE:AD:BE:EF:C0:DE", sname=b'', file=b'', options=dhcppython.options.OptionList([dhcppython.options.options.short_value_to_object(53, "DHCPDISCOVER")]))
>>> pkt
DHCPPacket(op='BOOTREQUEST', htype='ETHERNET', hlen=6, hops=0, xid=123456, secs=0, flags=0, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('0.0.0.0'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='DE:AD:BE:EF:C0:DE', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x01')]))
```

* Higher level constructors specific to the four main DHCP message types: DISCOVER, OFFER, REQUEST, ACK:

```python
>>> dhcppython.packet.DHCPPacket.Discover('de:ad:be:ef:c0:de')
DHCPPacket(op='BOOTREQUEST', htype='ETHERNET', hlen=6, hops=0, xid=4249353806, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('0.0.0.0'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='de:ad:be:ef:c0:de', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x01')]))
>>> dhcppython.packet.DHCPPacket.Offer('de:ad:be:ef:c0:de', seconds=0, tx_id=4249353806, yiaddr=ipaddress.IPv4Address('192.168.56.4'))
DHCPPacket(op='BOOTREPLY', htype='ETHERNET', hlen=6, hops=0, xid=4249353806, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('192.168.56.4'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='de:ad:be:ef:c0:de', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x02')]))
>>> dhcppython.packet.DHCPPacket.Request('de:ad:be:ef:c0:de', seconds=0, tx_id=4249353806)
DHCPPacket(op='BOOTREQUEST', htype='ETHERNET', hlen=6, hops=0, xid=4249353806, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('0.0.0.0'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='de:ad:be:ef:c0:de', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x03')]))
>>> dhcppython.packet.DHCPPacket.Ack('de:ad:be:ef:c0:de', seconds=0, tx_id=4249353806, yiaddr=ipaddress.IPv4Address('192.168.56.4'))
DHCPPacket(op='BOOTREPLY', htype='ETHERNET', hlen=6, hops=0, xid=4249353806, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('192.168.56.4'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='de:ad:be:ef:c0:de', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x05')]))
```

### dhcppython.options

This module provides classes for:

1. All DHCP options described in RFC 2132
2. An unknown option class for options not encoded
3. An abstract Option class that is easily extendable if additional options are required
4. A data structure for mananging DHCP options - the `OptionList`
5. An higher lever Option factory - the `OptionDirectory`

A high level API is provided by the `dhcppython.options.options` object and the Option class:

* Create an options object from bytes by calling the `bytes_to_object` method

```python
>>> opt = dhcppython.options.options.bytes_to_object(b"\x3d\x07\x01\x8c\x45\x00\x45\x12\x09")
>>> opt
ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00E\x12\t')
```

* Get a human readable dict of an options object value

```python
>>> opt.value
{'client_identifier': {'hwtype': 1, 'hwaddr': '8C:45:00:45:12:09'}}
```

* Create an options object from its human readable dict of its value:

```python
>>> dhcppython.options.options.value_to_object({'client_identifier': {'hwtype': 1, 'hwaddr': '8C:45:00:45:12:09'}})
ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00E\x12\t')
```

OR

```python
>>> dhcppython.options.options.short_value_to_object(61, {'hwtype': 1, 'hwaddr': '8C:45:00:45:12:09'})
ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00E\x12\t')
```

4. Convert a human readable dict of an option value to the bytes representation

```python
>>> dhcppython.options.options.value_to_bytes({'client_identifier': {'hwtype': 1, 'hwaddr': '8C:45:00:45:12:09'}})
b'=\x07\x01\x8cE\x00E\x12\t'
```

5. Get the bytes representation of an option given its Option object

```python
>>> opt = dhcppython.options.ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00E\x12\t')
>>> opt.asbytes
b'=\x07\x01\x8cE\x00E\x12\t'
```

The `OptionList` class provides a very convenient set of methods for managing a list of DHCP options.

* Create an `OptionList` instance from a list of `Option` objects

```python
>>> opt_list = dhcppython.options.OptionList(
...             [
...                 dhcppython.options.options.short_value_to_object(61, {'hwtype': 1, 'hwaddr': "8c:45:00:23:45:67"}),
...                 dhcppython.options.options.short_value_to_object(57, 1500),
...                 dhcppython.options.options.short_value_to_object(60, "android-dhcp-9"),
...                 dhcppython.options.options.short_value_to_object(12, "Galaxy-S9"),
...                 dhcppython.options.options.short_value_to_object(55, [1, 3, 6, 15, 26, 28, 51, 58, 59, 43])
...             ]
...         )
>>> opt_list
OptionList([ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00#Eg'), MaxDHCPMessageSize(code=57, length=2, data=b'\x05\xdc'), VendorClassIdentifier(code=60, length=14, data=b'android-dhcp-9'), Hostname(code=12, length=9, data=b'Galaxy-S9'), ParameterRequestList(code=55, length=10, data=b'\x01\x03\x06\x0f\x1a\x1c3:;+')])
```

* Retrieve any options in the `OptionList` by its option code

```python
>>> opt_list.by_code(12)
Hostname(code=12, length=9, data=b'Galaxy-S9')
>>> opt_list.by_code(13)
>>>
```

* Append (add) any options using the `append` method

```python
>>> opt_list.append(dhcppython.options.options.short_value_to_object(53, "DHCPDISCOVER"))
>>> opt_list
OptionList([ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00#Eg'), MaxDHCPMessageSize(code=57, length=2, data=b'\x05\xdc'), VendorClassIdentifier(code=60, length=14, data=b'android-dhcp-9'), Hostname(code=12, length=9, data=b'Galaxy-S9'), ParameterRequestList(code=55, length=10, data=b'\x01\x03\x06\x0f\x1a\x1c3:;+'), MessageType(code=53, length=1, data=b'\x01')])
```

* Protects against duplicate options (duplicate overwrites in place)

```python
>>> opt_list
OptionList([ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00#Eg'), MaxDHCPMessageSize(code=57, length=2, data=b'\x13\x88'), VendorClassIdentifier(code=60, length=14, data=b'android-dhcp-9'), Hostname(code=12, length=9, data=b'Galaxy-S9'), ParameterRequestList(code=55, length=10, data=b'\x01\x03\x06\x0f\x1a\x1c3:;+'), MessageType(code=53, length=1, data=b'\x01')])
```

* Allows for iteration like a list

```python
>>> for opt in opt_list:
...     print(opt)
...
ClientIdentifier(code=61, length=7, data=b'\x01\x8cE\x00#Eg')
MaxDHCPMessageSize(code=57, length=2, data=b'\x13\x88')
VendorClassIdentifier(code=60, length=14, data=b'android-dhcp-9')
Hostname(code=12, length=9, data=b'Galaxy-S9')
ParameterRequestList(code=55, length=10, data=b'\x01\x03\x06\x0f\x1a\x1c3:;+')
MessageType(code=53, length=1, data=b'\x01')
```

## The DHCP Client

A very primitive DHCP client is included in this package in the `dhcppython.client` module. The client is able to negotiate a lease with a DHCP server and can be configured to use:

* A given interface
* Option to send broadcast packets or unicast packets to a specific server
* Set a relay in the giaddr field
* "Spoof" MAC addresses
* Specify options to send with request

The high level interface to negotiate a lease is the `get_lease` method of the `dhcppython.client.DHCPClient` object. This method goes through the DORA DHCP handshake and returns a `Lease` namedtuple which includes all the packets in the :

```python
>>> import dhcppython
>>> client = dhcppython.client.DHCPClient(interface="enp0s8")
>>> lease = client.get_lease(mac_addr="de:ad:be:ef:c0:de", broadcast=True, relay=None, server="255.255.255.255", options_list=None)
Lease succesful: 192.168.56.3 -- DE:AD:BE:EF:C0:DE -- 3 ms elapsed
>>> lease
Lease(discover=DHCPPacket(op='BOOTREQUEST', htype='ETHERNET', hlen=6, hops=0, xid=2829179566, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('0.0.0.0'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='de:ad:be:ef:c0:de', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x01')])), offer=DHCPPacket(op='BOOTREPLY', htype='ETHERNET', hlen=6, hops=0, xid=2829179566, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('192.168.56.3'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='DE:AD:BE:EF:C0:DE', sname=b'', file=b'', options=OptionList([SubnetMask(code=1, length=4, data=b'\xff\xff\xff\x00'), Router(code=3, length=4, data=b'\n\x97\x01\x01'), DNSServer(code=6, length=4, data=b'\nh\x01\x08'), Hostname(code=12, length=22, data=b'dhcp.-192-168-56-3.com'), DomainName(code=15, length=14, data=b'example.com'), IPAddressLeaseTime(code=51, length=4, data=b'\x00\x01Q\x80'), MessageType(code=53, length=1, data=b'\x02'), ServerIdentifier(code=54, length=4, data=b'\xc0\xa88\x02'), RenewalTime(code=58, length=4, data=b'\x00\x00T`'), RebindingTime(code=59, length=4, data=b'\x00\x00\xa8\xc0'), End(code=255, length=0, data=b'')])), request=DHCPPacket(op='BOOTREQUEST', htype='ETHERNET', hlen=6, hops=0, xid=2829179566, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('0.0.0.0'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='de:ad:be:ef:c0:de', sname=b'', file=b'', options=OptionList([MessageType(code=53, length=1, data=b'\x03')])), ack=DHCPPacket(op='BOOTREPLY', htype='ETHERNET', hlen=6, hops=0, xid=2829179566, secs=0, flags=32768, ciaddr=IPv4Address('0.0.0.0'), yiaddr=IPv4Address('192.168.56.3'), siaddr=IPv4Address('0.0.0.0'), giaddr=IPv4Address('0.0.0.0'), chaddr='DE:AD:BE:EF:C0:DE', sname=b'', file=b'', options=OptionList([SubnetMask(code=1, length=4, data=b'\xff\xff\xff\x00'), Router(code=3, length=4, data=b'\n\x97\x01\x01'), DNSServer(code=6, length=4, data=b'\nh\x01\x08'), Hostname(code=12, length=22, data=b'dhcp.-192-168-56-3.com'), DomainName(code=15, length=14, data=b'example.com'), IPAddressLeaseTime(code=51, length=4, data=b'\x00\x01Q\x80'), MessageType(code=53, length=1, data=b'\x05'), ServerIdentifier(code=54, length=4, data=b'\xc0\xa88\x02'), RenewalTime(code=58, length=4, data=b'\x00\x00T`'), RebindingTime(code=59, length=4, data=b'\x00\x00\xa8\xc0'), End(code=255, length=0, data=b'')])), time=0.0032514659978915006, server=('192.168.56.2', 67))
```


CHANGELOG.md
# DHCP Python Changelog

## 0.1.4 (Mar 30 2021)

* Merges (PR #6)[https://github.com/vfrazao-ns1/dhcppython/pull/6] fixing a bug where MAC addresses were stripped from of 0 bytes from both ends


## 0.1.3 (Aug 19 2020)

* Merges (PR #4)[https://github.com/vfrazao-ns1/dhcppython/pull/4] which extends option 82 support

## 0.1.2 (Jan 29 2020)

* Minor fixes and changes

## 0.1.1 (Jan 25 2020)

* Minor bug fixes
* Code formatting with black

## 0.1.0 (Jan 25 2020)

* Initial release

dhcppython/utils.py
import random
import string
import datetime
import socket
import unicodedata
from typing import Dict
import importlib.resources
from . import runtime_assets


VALID_HEX = list(set(string.hexdigits.upper()))


def cur_datetime(us_precision: bool = False) -> str:
    fmt = "%Y-%m-%dT%H:%M:%S" + (".%f" if us_precision else "") + "Z"
    return datetime.datetime.now(datetime.timezone.utc).strftime(fmt)


def cur_timestamp() -> int:
    return int(datetime.datetime.utcnow().timestamp() * 10 ** 9)


def visual_length(text: str) -> int:
    """
    Given a string it returns the visual length of the string as opposed to the
    len function which returns the number of printable characters.
    """
    # See https://www.unicode.org/reports/tr11/ for how this dict in constructed
    visual_len = {
        "F": 1,
        "H": 1,
        "Na": 1,
        "N": 1,
        "W": 2,
        "A": 2,
    }
    return sum([visual_len[unicodedata.east_asian_width(char)] for char in text]) + 1


def random_mac(num_bytes: int = 6, delimiter: str = ":") -> str:
    """
    Generates an 6 byte long MAC address.

    >>> random_mac()
    'CC:AC:3C:85:A4:EF'
    """
    return delimiter.join(
        ["".join(random.choices(VALID_HEX, k=2)) for i in range(num_bytes)]
    )


def is_mac_addr(mac_addr: str) -> bool:
    """
    Returns True if the string is a valid MAC address.

    Accepts ":" or "-" as valid MAC address delimiters.
    """
    mac_addr = mac_addr.upper()
    delimiter = ":" if ":" in mac_addr else "-"
    if len(mac_addr.split(delimiter)) != 6 or len(mac_addr) != 17:
        return False
    if any([b not in VALID_HEX for b in "".join(mac_addr.split(delimiter))]):
        return False
    return True

try:
    oui_lines = importlib.resources.read_text(runtime_assets, "oui.txt").split("\n")
except:
    oui_lines = []
mac_vendor_map: Dict[str, str] = {
    line.split("\t\t")[0].split(" ")[0]: line.split("\t\t")[1]
    for line in [
        line.strip()
        for line in oui_lines
        if "(base 16)" in line
    ]
}


def mac2vendor(mac_addr: str) -> str:
    if is_mac_addr(mac_addr):
        return mac_vendor_map.get(
            mac_addr.replace(":", "").replace("-", "")[:6].upper(),
            "Unknown Manufacturer",
        )
    else:
        raise ValueError(f"{mac_addr} is not a valid MAC address")


def get_ip_by_iface(iface: str) -> str:
    rand_port = 61224
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.setsockopt(socket.SOL_SOCKET, 25, iface.encode())
    s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    s.connect(("255.255.255.255", rand_port))
    return s.getsockname()[0]


def get_ip_by_server(server: str) -> str:
    rand_port = 61222
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect((server, rand_port))
    return s.getsockname()[0]


dhcppython/dora.py
import logging
import argparse
import json
from dhcppython import client, options, utils


if __name__ == "__main__":
    logging.basicConfig(
        level=logging.DEBUG,
        format="%(asctime)s - %(threadName)s - %(levelname)s: %(message)s",
        handlers=[logging.StreamHandler()],
    )

    help_texts = {
        "main": __doc__,
        "interface": "Interface to bind to and make DHCP requests",
        "mac_addr": "MAC address to use (default random)",
        "debug": "Print debug statements",
        "unicast": "Send DHCP packets over unicast to specified server",
        "server": "Server to send DHCP packets. Required for unicast and for relay use.",
        "relay": "Address to set the giaddr field to",
        "verbose": "Send random IP address",
        "options": "JSON body of options to include in requests",
        "port": "Port to send packets from on client machine",
        "target_port": "Port to send to on target machine",
        "target": "Given an IP address of a DHCP server, sends unicast requests",
    }
    parser = argparse.ArgumentParser(
        description=help_texts["main"], formatter_class=argparse.RawTextHelpFormatter,
    )

    parser.add_argument("-i", "--interface", type=str, help=help_texts["interface"])
    parser.add_argument("-a", "--mac_addr", type=str, default=utils.random_mac(), help=help_texts["mac_addr"])
    parser.add_argument("-d", "--debug", action="store_true", help=help_texts["debug"])
    parser.add_argument("-u", "--unicast", action="store_true", help=help_texts["unicast"])
    parser.add_argument("-s", "--server", type=str, default="255.255.255.255", help=help_texts["server"])
    parser.add_argument("-r", "--relay", type=str, help=help_texts["relay"])
    parser.add_argument('-v', '--verbose', action='count', default=0, help=help_texts["verbose"])
    parser.add_argument("-o", "--options", type=str, help=help_texts["options"])
    parser.add_argument("-p", "--port", type=int, default=68, help=help_texts["port"])
    parser.add_argument("--target_port", type=int, default=67, help=help_texts["target_port"])
    parser.add_argument("-@", type=str, dest="target", help=help_texts["target"])

    args = parser.parse_args()

    debug = args.debug
    if args.verbose > 2:
        debug = True

    if debug:
        print(args)

    if not debug:
        logging.disable(logging.DEBUG)

    mac_addr = args.mac_addr
    broadcast = not(args.unicast)
    server = args.server
    relay = args.relay
    verbosity = args.verbose

    if args.target:
        broadcast = False
        server = args.target
        relay = utils.get_ip_by_server(server)

    # Default options set
    opts = [
        options.options.short_value_to_object(61, {'hwtype': 1, 'hwaddr': mac_addr}),
        options.options.short_value_to_object(57, 1500),
        options.options.short_value_to_object(60, "android-dhcp-9"),
        options.options.short_value_to_object(12, "Galaxy-S9"),
        options.options.short_value_to_object(55, [1, 3, 6, 15, 26, 28, 51, 58, 59, 43])
    ]
    if args.options:
        try:
            with open(args.options) as opt_file:
                opt_json = opt_file.read()
                opts_dict = json.loads(opt_json)
                opts = []
                for k,v in opts_dict.items():
                    opts.append(options.options.value_to_object({k: v}))
        except Exception as e:
            logging.error(f"Unable to parse JSON options file {args.option}: {e}")
            exit(1)
    options_list = options.OptionList(opts)
    if client_addr := options_list.by_code(61):
        mac_addr = client_addr.value["client_identifier"]["hwaddr"]

    if verbosity:
        print(f"{client.CLIENT_NAME} - v{client.CLIENT_VER} - {utils.cur_datetime()}")
    c = client.DHCPClient(args.interface, send_from_port=args.port, send_to_port=args.target_port)
    try:
        lease = c.get_lease(mac_addr, broadcast=broadcast, options_list=options_list, server=server, relay=relay, verbose=verbosity)
    except Exception as e:
        logging.debug(e)
        logging.error("Unable to obtain lease, run with -d to debug")
        if debug:
            raise
